compiler/rustc_parse/src/errors.rs

```diff
@@ -433,6 +433,18 @@ pub(crate) enum MissingInInForLoopSub {
     AddIn(#[primary_span] Span),
 }
 
+#[derive(Diagnostic)]
+#[diag(parse_missing_expression_in_for_loop)]
+pub(crate) struct MissingExpressionInForLoop {
+    #[primary_span]
+    #[suggestion(
+        code = "/* expression */ ",
+        applicability = "machine-applicable",
```

Comment:
This suggestion is not machine-applicable, because the suggestion is not valid syntax after it's applied. It's just there to help the user understand the issue.

```suggestion
        applicability = "has-placeholders",
```